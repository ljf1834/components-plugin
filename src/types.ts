import type { FilterPattern, ViteDevServer } from "vite"

export type SupportedTransformer = 'vue3' | 'vue2'
export interface Options {
  /**
   * RegExp or glob to match files to be transformed
   */
  include?: FilterPattern

  /**
   * RegExp or glob to match files to NOT be transformed
   */
  exclude?: FilterPattern
  /**
   * Relative paths to the directory to search for components.
   * @default 'src/components'
   */
  dirs?: string | string[]

  /**
   * Valid file extensions for components.
   * @default ['vue']
   */
  extensions?: string | string[]

  /**
   * Glob patterns to match file names to be detected as components.
   *
   * When specified, the `dirs` and `extensions` options will be ignored.
   */
  globs?: string | string[]

  /**
   * Search for subdirectories
   * @default true
   */
  deep?: boolean

  /**
   * Subdirectory paths for ignoring namespace prefixes
   *
   * Works when `directoryAsNamespace: true`
   * @default "[]"
   */
  globalNamespaces?: string[]

  /**
   * Collapse same prefixes (case-insensitive) of folders and components
   * to prevent duplication inside namespaced component name
   *
   * Works when `directoryAsNamespace: true`
   * @default false
   */
  collapseSamePrefixes?: boolean

  /**
   * Allow subdirectories as namespace prefix for components
   * @default false
   */
  directoryAsNamespace?: boolean

  /**
   * Apply custom transform over the path for importing
   */
  importPathTransform?: (path: string) => string | undefined

  /**
   * Transformer to apply
   *
   * @default 'vue3'
   */
  transformer?: SupportedTransformer
  /**
   * Generate TypeScript declaration for global components
   *
   * Accept boolean or a path related to project root
   *
   * @see https://github.com/vuejs/core/pull/3399
   * @see https://github.com/johnsoncodehk/volar#using
   * @default true
   */
  dts?: boolean | string
}
export type ResolveOptions = Omit<Options, 'extensions' | 'dirs'>  & {
  extensions: string[]
  dirs: string[]
  resolvedDirs: string[]
  dts?: string | false
  root: string
  globs: string[]
}
export type ResolveResult = {
  rawName: string
  replace: (resolved: string) => void
}
export type Nullable<T> = T | null | undefined
export type Arrayable<T> = T | Array<T>
export interface interfaceContext {
  root: string,
  options: ResolveOptions,
  alias: Record<string, string>,
  _server: ViteDevServer | undefined,
  _globs: string[],
  _componentPaths: Set<string>,
  _componentNamesMap: Record<string, string>,
  setRoot(root: string): void,
  setServer(server: ViteDevServer): void,
  findComponent(name: string): void,
  searchComponents(): void,
  resolveOptions(rawOptions: Options, root: string): ResolveOptions,
  resolveGlobs(dirs: string[]): string[],
  transform(code: string, id: string): { code: string, map: any },
  get componentNameMap(): Record<string, string>,
  get globs(): string[],
}
