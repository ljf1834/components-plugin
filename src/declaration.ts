import type { Context } from "./context"

export function writeDeclaration(ctx: Context, filePath: string) {

}

function generateDeclarantCode(ctx: Context) {

}
