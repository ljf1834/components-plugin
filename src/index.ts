import { type PluginOption, type ResolvedConfig } from 'vite'
import { resolve } from 'node:path'
import fg from 'fast-glob'

export function createComponentPlugin():PluginOption {
  let root = process.cwd()
  const dirs = 'src/components'
  const componentsMap = {}
  function findComponents(glob:(string)[]){
    const paths = fg.sync(glob, { cwd: root })
    console.log(paths)
  }
  function resolveGlob() {

  }

  return {
    name: 'component-plugin',
    configResolved(config: ResolvedConfig) {
      root = config.root
      findComponents([dirs])
      console.log('---', root)
    },
    transform: (code, id) => {

    }
  }
}
