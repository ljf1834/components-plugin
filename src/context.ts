import {getNameFromFilePath, parseId, pascalCase, slash, stringifyComponentImport, toArray, matchGlobs} from "./utils"
import {resolve} from "node:path"
import fg from 'fast-glob'
import MagicString from 'magic-string'
import { ViteDevServer } from "vite"
import {fileURLToPath, pathToFileURL, URL} from "node:url"

export interface Options {
  /**
   * Relative paths to the directory to search for components.
   * @default 'src/components'
   */
  dirs?: string | string[]

  /**
   * Valid file extensions for components.
   * @default ['vue']
   */
  extensions?: string | string[]

  /**
   * Glob patterns to match file names to be detected as components.
   *
   * When specified, the `dirs` and `extensions` options will be ignored.
   */
  globs?: string | string[]

  /**
   * Search for subdirectories
   * @default true
   */
  deep?: boolean

  /**
   * Subdirectory paths for ignoring namespace prefixes
   *
   * Works when `directoryAsNamespace: true`
   * @default "[]"
   */
  globalNamespaces?: string[]

  /**
   * Collapse same prefixes (case-insensitive) of folders and components
   * to prevent duplication inside namespaced component name
   *
   * Works when `directoryAsNamespace: true`
   * @default false
   */
  collapseSamePrefixes?: boolean

  /**
   * Allow subdirectories as namespace prefix for components
   * @default false
   */
  directoryAsNamespace?: boolean

  /**
   * Apply custom transform over the path for importing
   */
  importPathTransform?: (path: string) => string | undefined
}
export type ResolveOptions = Omit<Options, 'extensions' | 'dirs'>  & {
  extensions: string[]
  dirs: string[]
  resolvedDirs: string[]
  dts?: string | false
  root: string
  globs: string[]
}
interface ResolveResult {
  rawName: string
  replace: (resolved: string) => void
}
const defaultOptions = {
  dirs: 'src/components',
  extensions: 'vue',
  deep: true,
  importPathTransform: v => v
}

export class Context {
  root = process.cwd()
  alias: Record<string, string>
  options: ResolveOptions
  private _globs:string[] = []
  private _componentPaths: Set<string> = new Set()
  private _componentNamesMap = {}
  constructor(private rawOptions: Options = {}) {
    this.options = this.resolveOptions(rawOptions, this.root)
    this._globs = this.resolveGlobs(this.options.resolvedDirs)
  }
  setRoot(root: string) {
    this.root = root
    this.options = this.resolveOptions(this.rawOptions, this.root)
    this._globs = this.resolveGlobs(this.options.dirs)
  }
  private server:ViteDevServer | undefined
  setServer(server: ViteDevServer) {
    this.server = server
    this.server.watcher.on('add', (filePath) => {
      if (!matchGlobs(filePath, this.options.dirs.map(path => fileURLToPath(new URL(path, pathToFileURL(this.root).toString() + '/').href)))) return
      this.searchComponents()
    })
    this.server.watcher.on('unlink', (filePath) => {
      if (!matchGlobs(filePath, this.options.dirs.map(path => fileURLToPath(new URL(path, pathToFileURL(this.root).toString() + '/').href)))) return
      this.searchComponents()
    })
  }
  resolveOptions(rawOptions: Options, root):ResolveOptions{
    const resolveOptions = Object.assign({}, defaultOptions, rawOptions) as ResolveOptions
    resolveOptions.extensions = toArray(resolveOptions.extensions)
    resolveOptions.dirs = toArray(resolveOptions.dirs)
    resolveOptions.resolvedDirs = resolveOptions.dirs.map(u => slash(resolve(root, u)))
    resolveOptions.root = root
    return resolveOptions
  }
  resolveGlobs(dirs:string[]) {
    if (this.options.globs) {
      return toArray(this.options.globs).map((glob: string) => slash(resolve(this.options.root, glob)))
    } else {
      const extsGlob = this.options.extensions.length==1 ? this.options.extensions[0] : `{${(this.options.extensions as string[]).join(',')}}`
      return dirs.map((dir: string) => this.options.deep ? slash(`./${dir}/**/*.${extsGlob}`) : slash(`./${dir}/*.${extsGlob}`))
      if (!this.options.extensions.length) throw new Error('[components-plugin] `extensions` option is required to search for components')
    }
  }
  searchComponents() {
    const paths = fg.sync(this._globs, { cwd: this.options.root, ignore: ['node_modules'], onlyFiles: true, absolute: true})
    toArray(paths).forEach(item => this._componentPaths.add(item))
    Array.from(this._componentPaths).forEach((path) => {
      const name = pascalCase(getNameFromFilePath(path, this.options))
      this._componentNamesMap[name] = {
        as: name,
        from: path
      }
    })
  }
  findComponent(name) {
    return this._componentNamesMap[name]
  }
  transform(code: string, id: string) {
    let no = 0
    const { path, query } = parseId(id)
    const s = new MagicString(code)
    const results: ResolveResult[] = []
    for (const match of code.matchAll(/_resolveComponent[0-9]*\("(.+?)"\)/g)) {
      const matchedName = match[1]
      if (match.index != null && matchedName && !matchedName.startsWith('_')) {
        const start = match.index
        const end = start + match[0].length
        results.push({
          rawName: matchedName,
          replace: resolved => s.overwrite(start, end, resolved),
        })
      }
    }
    for (const { rawName, replace } of results) {
      const name = pascalCase(rawName)
      const component = this.findComponent(name)
      if (component) {
        const varName = `components_${no}`
        s.prepend(`${stringifyComponentImport({ ...component, as: varName }, this)};\n`)
        no += 1
        replace(varName)
      }
    }
    return {
      code: s.toString(),
      map: s.generateMap({ source: id, includeContent: true })
    }
  }
  get componentNameMap() {
    return this._componentNamesMap
  }
  get globs() {
    return this._globs
  }
}
