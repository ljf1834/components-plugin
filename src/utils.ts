import fg from "fast-glob";
import { parse } from 'path'
import { type ResolveOptions, type Options } from "./context"

type Nullable<T> = T | null | undefined
type Arrayable<T> = T | Array<T>

export function toArray<T>(array?: Nullable<Arrayable<T>>): Array<T> {
  array = array ?? []
  return Array.isArray(array) ? array : [array]
}

export function slash(str: string) {
  return str.replace(/\\/g, '/')
}

export function isEmpty(value: any) {
  if (!value || value === null || value === undefined || (Array.isArray(value) && Object.keys(value).length <= 0))
    return true
  else
    return false
}

export function getNameFromFilePath(filePath: string, options: ResolveOptions): string {
  const { resolvedDirs, directoryAsNamespace, globalNamespaces, collapseSamePrefixes, root } = options
  const parsedFilePath = parse(slash(filePath))
  let strippedPath = ''
  // remove include directories from filepath
  for (const dir of resolvedDirs) {
    if (parsedFilePath.dir.startsWith(dir)) {
      strippedPath = parsedFilePath.dir.slice(dir.length)
      break
    }
  }
  let folders = strippedPath.slice(1).split('/').filter(Boolean)
  let filename = parsedFilePath.name
  // set parent directory as filename if it is index
  if (filename === 'index' && !directoryAsNamespace) {
    // when use `globs` option, `resolvedDirs` will always empty, and `folders` will also empty
    if (isEmpty(folders))
      folders = parsedFilePath.dir.slice(root.length + 1).split('/').filter(Boolean)
    filename = `${folders.slice(-1)[0]}`
    return filename
  }
  if (directoryAsNamespace) {
    // remove namesspaces from folder names
    if (globalNamespaces.some((name: string) => folders.includes(name)))
      folders = folders.filter(f => !globalNamespaces.includes(f))
    folders = folders.map(f => f.replace(/[^a-zA-Z0-9\-]/g, ''))
    if (filename.toLowerCase() === 'index')
      filename = ''
    if (!isEmpty(folders)) {
      // add folders to filename
      let namespaced = [...folders, filename]
      if (collapseSamePrefixes) {
        const collapsed: string[] = []
        for (const fileOrFolderName of namespaced) {
          const collapsedFilename = collapsed.join('')
          if (collapsedFilename && fileOrFolderName.toLowerCase().startsWith(collapsedFilename.toLowerCase())) {
            const collapseSamePrefix = fileOrFolderName.slice(collapsedFilename.length)
            collapsed.push(collapseSamePrefix)
            continue
          }
          collapsed.push(fileOrFolderName)
        }
        namespaced = collapsed
      }
      filename = namespaced.filter(Boolean).join('-')
    }
    return filename
  }
  return filename
}

export function pascalCase(str: string) {
  return capitalize(camelCase(str))
}

export function camelCase(str: string) {
  return str.replace(/-(\w)/g, (_, c) => (c ? c.toUpperCase() : ''))
}

export function kebabCase(key: string) {
  const result = key.replace(/([A-Z])/g, ' $1').trim()
  return result.split(' ').join('-').toLowerCase()
}

export function capitalize(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

export function parseId(id: string) {
  const index = id.indexOf('?')
  if (index < 0) {
    return { path: id, query: {} }
  }
  else {
    const query = Object.fromEntries(new URLSearchParams(id.slice(index)) as any)
    return {
      path: id.slice(0, index),
      query,
    }
  }
}

export function stringifyComponentImport({ as: name, from: path, name: importName, sideEffects }, ctx) {
  path = getTransformedPath(path, ctx.options.importPathTransform)

  const imports = [
    stringifyImport({ as: name, from: path, name: importName }),
  ]

  if (sideEffects)
    toArray(sideEffects).forEach(i => imports.push(stringifyImport(i)))

  return imports.join(';')
}

export function getTransformedPath(path: string, importPathTransform?: Options['importPathTransform']): string {
  if (importPathTransform) {
    const result = importPathTransform(path)
    if (result != null)
      path = result
  }
  return path
}

export function stringifyImport(info) {
  if (typeof info === 'string')
    return `import '${info}'`
  if (!info.as)
    return `import '${info.from}'`
  else if (info.name)
    return `import { ${info.name} as ${info.as} } from '${info.from}'`
  else
    return `import ${info.as} from '${info.from}'`
}
