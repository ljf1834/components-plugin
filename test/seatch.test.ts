import { describe, expect, it } from 'vitest'
import { Context } from '../src/context'
import { resolve } from "node:path"

const root = resolve(__dirname, '../examples/vue3')

describe('search', () => {
  it('should work', async () => {
    const ctx = new Context({})
    ctx.setRoot(root)
    ctx.searchComponents()

    expect(ctx.componentNameMap).toMatchSnapshot()
  })
})
