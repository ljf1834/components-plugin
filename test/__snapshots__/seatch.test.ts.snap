// Vitest Snapshot v1

exports[`search > should work 1`] = `{}`;
